### ISSUE TITLE
/search.rss and /search?format=rss returns HTTP 500

### ISSUE THREAD

**Issue by da2x** [lobsters/lobsters#746]
The following links return a HTTP 500. I realize that RSS from search results isn't an officially documented and supported feature, but the two first URLs should probably return the HTML page instead if RSS isn't supported.

https://lobste.rs/search?order=newest&q=domain:arstechnica.com&format=rss
https://lobste.rs/search?order=newest&q=interesting&format=rss
https://lobste.rs/search.rss?order=newest&q=domain:arstechnica.com

Update: also applies to &format=atom and search.atom.

**Comment by finnoleary** [comment]
Could you give an example of a search that it does work on?

Also, wouldn't it be correct functionality, though? I do not know the conventions around Atom or RSS, but trying to parse HTML as RSS will fail, which is likely to do unexpected things depending on how each feed reader chooses to handle it. Whereas leaving it as a 500 makes any possible feed reader behaviour entirely predictable. 

I guess what I'm really asking is, what's your use case, and is this normal expected behaviour for RSS feeds?

**Comment by pushcx** [comment]
Yeah. We should probably fail with 400 (bad request).

**Comment by pushcx** [comment]
I merged #765 to address this and noticed there's a different exception if the action uses render without respond_to and added a couple tests, one of which is not behaving: 6c804ee

**Comment by emrox** [comment]
yes, I'll have to adapt it. Hope I can find some time this week

**Comment by pushcx** [comment]
I spent a while tinkering with this. If there's a `render` call that named an `action` or `template`, but no `format` block, Rails tries to look up a template for any format requested and fails, throwing `MissingTemplate`. The implict render doesn't throw. This difference is weird and possibly buggy.

In 891aabbf I removed the `render` calls that exhibit this behavior and are redundant because they give the name of the action, which the implict render finds anyways. This still leaves the problem for `/recent` because it renders the `index` template. So, solutions?

We could add `:constraints => { :format => [:html]}` to nearly every route, so much for convention over configuration.

We could add a `format` block to every `render`, so much for convention over configuration.

We could wrap `render` in `ApplicationController` to add a `format` block if not present. It sounds straightforward, but `render` has a complicated interface so it feels likely this would get bitten by unknowns.

We could symlink templates, e.g. from `recent.html.erb` to `index.html.erb`, but that seems like a setup to confuse someone who doesn't notice the templates are all the same file, or renames a template and breaks the symlinks.

We could rescue `MissingTemplate` in `ApplicationController`. This would swallow real exceptions when templates are actually missing, but if it passed through exceptions when the requested format was html, that's probably not an impossible-to-debug time bomb. This feels pretty promising.

We could add a `constraints ... do` or `scope` block around the routes, but routes that support rss or json are sprinkled throughout our routes. And digging through our history, we _tried_ that: in 2014, we added in PR #116 to fix bug #114 - which this bug is a dupe of! Not knowing why it was there, I removed it in #482.  Aaaand reading that, `/search.asdf` with the implicit render happily renders an html page, which is the second weird, possibly buggy behavior by Rails in this comment.

So now that the odd behavior is exposed, maybe the best fix is to restore the `scope` around all of the routes and force routes that support other formats to explicitly name them. Gonna sleep on this one a minute, but if anyone wants to research if Rails considers either of the odd behaviors to be bugs, that'd help a lot.
